token-swap/program/src/processor.rs

```diff
@@ -163,28 +163,42 @@ impl Processor {
         }
         let token_a = Self::unpack_token_account(&token_a_info.data.borrow())?;
         let token_b = Self::unpack_token_account(&token_b_info.data.borrow())?;
+        let destination = Self::unpack_token_account(&destination_info.data.borrow())?;
         let pool_mint = Self::unpack_mint(&pool_info.data.borrow())?;
         if *authority_info.key != token_a.owner {
             return Err(SwapError::InvalidOwner.into());
         }
         if *authority_info.key != token_b.owner {
             return Err(SwapError::InvalidOwner.into());
         }
-        if spl_token::option::COption::Some(*authority_info.key) != pool_mint.mint_authority {
+        if *authority_info.key == destination.owner {
+            return Err(SwapError::InvalidOutputOwner.into());
+        }
+        if COption::Some(*authority_info.key) != pool_mint.mint_authority {
             return Err(SwapError::InvalidOwner.into());
         }
+        if token_a.mint == token_b.mint {
+            return Err(SwapError::RepeatedMint.into());
+        }
         if token_b.amount == 0 {
-            return Err(SwapError::InvalidSupply.into());
+            return Err(SwapError::EmptySupply.into());
         }
         if token_a.amount == 0 {
-            return Err(SwapError::InvalidSupply.into());
+            return Err(SwapError::EmptySupply.into());
         }
         if token_a.delegate.is_some() {
             return Err(SwapError::InvalidDelegate.into());
         }
         if token_b.delegate.is_some() {
             return Err(SwapError::InvalidDelegate.into());
         }
+        if pool_mint.supply != 0 {
+            return Err(SwapError::InvalidSupply.into());
+        }
+        // double check to be totally sure -- TODO check if this is necessary
```

Comment:
Is it worth having this second check?  I'm hoping that just checking the pool mint supply is enough.

Reply:
Great idea -- I'll add something to SwapInstruction::Initialize then